torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -20,3 +68,137 @@ def __init__(
+                )
+            self._global_mesh = self.mesh_info.mesh
+            self._global_placements = (Shard(0),)
+            self._global_size = param.size()
+            self._global_stride = param.stride()
+            param_data = param
+        self._orig_size = param_data.size()
+        shard_rank = self.mesh_info.shard_mesh_rank
+        shard_world_size = self.mesh_info.shard_mesh_size
+        chunks = _chunk_with_empty(param_data, shard_world_size, dim=0)
+        sharded_param = chunks[shard_rank]
+        self.sharded_size = sharded_param.size()
+        padded_sharded_size = chunks[0].size()  # 0th always padded
+        padded_sharded_param = param_data.new_zeros(padded_sharded_size)
+        if sharded_param.numel() > 0:
+            padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
+        self._sharded_param_data = padded_sharded_param.view(-1)
+        self.sharded_param = nn.Parameter(
+            self.to_sharded_dtensor(padded_sharded_param[: sharded_param.size(0)])
+        )
+        self.sharded_param.requires_grad_(param.requires_grad)
+        unsafe_free_storage(param_data)  # free immediately
+        del param_data  # delete PyObject reference to avoid warning
+        self._setattr_on_modules(self.sharded_param)
+        self.sharded_state = ShardedState.SHARDED
+
+    def _setattr_on_modules(self, tensor: torch.Tensor) -> None:
+        unsafe_setattr_param(
+            self._module_info.module, self._module_info.param_name, tensor
+        )
+        for shared_module, shared_param_name in zip(
+            self._module_info.shared_modules, self._module_info.shared_param_names
+        ):
+            unsafe_setattr_param(shared_module, shared_param_name, tensor)
+
+    def to_sharded_dtensor(self, tensor: torch.Tensor) -> DTensor:
+        """
+        Converts a local tensor representing either the sharded parameter or
+        sharded gradient to DTensor.
+        """
+        if tensor.shape != self.sharded_size and not (
+            # For size-0 padding, DTensor can flatten from (0, *) to (0)
+            tensor.numel() == 0
+            and self.sharded_size.numel() == 0
+        ):
+            _raise_assert_with_print(
+                f"Expects a tensor with the sharded size {self.sharded_size} "
+                f"but got {tensor.shape}"
+            )
+        return _from_local_no_grad(
+            tensor,
+            self._global_mesh,
+            self._global_placements,
+            self._global_size,
+            self._global_stride,
+        )
+
+
+# NOTE: Unsafe here refers to not checking whether the storage is already
+# allocated or freed, respectively. We should be safe to use them since we
+# explicitly manage the state transition.
+def unsafe_alloc_storage(tensor: torch.Tensor) -> None:
+    # Skip the already-allocated check and assume that `tensor` is the base
+    # tensor to save CPU overhead
```

Comment:
are we indicating high cpu times by using
```storage_size: int = tensor.untyped_storage().size()``` ? OR will it trigger cuda async? 

Trying to understand if this is chasing for perf savings or code elegance

Reply:
This is all CPU side overhead. You are correct that it is saving the `tensor.untyped_storage().size()` check(s). E.g., we could check if the size is already the allocated size and not do anything, and we could check if the size is greater than zero but not the desired size and raise an error. We could also check that the tensor is the base tensor / is not a view with nonzero storage offset.

I mainly avoided that because it was not necessary for our use case. I did not want this `unsafe_alloc_storage` to be a general function that others (outside FSDP) would use. If you are using this, you should have a reason for it. For our design, we make sure that we will never call it when the storage size is greater than 0 by guarding with the `ShardedState` state transitions, and we never call it when the tensor is not the base tensor since FSDP manages the tensor that we are passing in.

If people prefer, I can add back those checks because they will not incur too much CPU overhead, so it does not matter much either way.